query/transformation.go

```diff
@@ -0,0 +1,156 @@
+			scale:     scale,
+			parameter: parameter,
+		})
+	}
+	return result
+}
+
+// transformDerivative estimates the "change per second" between the two samples (scaled consecutive difference)
+func transformDerivative(values []float64, parameter transformationParameter) []float64 {
+	result := make([]float64, len(values))
+	for i := range values {
+		if i == 0 {
+			// The first element has 0
+			result[i] = 0
+			continue
+		}
+		// Otherwise, it's the scaled difference
+		result[i] = (values[i] - values[i-1]) / parameter.scale
+	}
+	return result
+}
+
+// transformIntegral integrates a series whose values are "X per second" to estimate "total X so far"
+func transformIntegral(values []float64, parameter transformationParameter) []float64 {
+	result := make([]float64, len(values))
+	integral := 0.0
+	for i := range values {
+		integral += values[i]
+		result[i] = integral * parameter.scale
+	}
+	return result
+}
+
+// transformRate functions exactly like transformDerivative but bounds the result to be positive and does not normalize
+func transformRate(values []float64, parameter transformationParameter) []float64 {
+	result := make([]float64, len(values))
+	for i := range values {
+		if i == 0 {
+			result[i] = 0
+			continue
+		}
+		result[i] = (values[i] - values[i-1]) / parameter.scale
+		if result[i] < 0 {
+			result[i] = 0
+		}
+	}
+	return result
+}
+
+// transformMovingAverage finds the average over the time period given in the parameter.parameter value
+func transformMovingAverage(values []float64, parameter transformationParameter) []float64 {
+	result := make([]float64, len(values))
+	limit := int(parameter.parameter/parameter.scale + 0.5) // Limit is the number of items to include in the average
+	if limit < 1 {
+		// At least one value must be included at all times
+		limit = 1
+	}
+	count := 0
+	sum := 0.0
+	for i := range values {
+		sum += values[i]
+		if count < limit {
+			// Increment the number of participants.
+			count++
+		} else {
+			// Remove the earliest participant
+			// (This is an optimization)
+			sum -= values[i-limit]
+			// For example, if limit = 4, we are suppose to include 4 items (including oneself).
+			// If i = 10, then we want to include 10, 9, 8, 7.
+			// So exclude 6 = 10 - 4 = i - limit
+		}
+		result[i] = sum / float64(count)
+	}
+	return result
+}
+
+// transformMapMaker can be used to use a function as a transformation, such as 'math.Abs' (or similar):
+//  `transformMapMaker(math.Abs)` is a transformation function which can be used, e.g. with applyTransformation
+func transformMapMaker(fun func(float64) float64) func([]float64, transformationParameter) []float64 {
+	return func(values []float64, parameter transformationParameter) []float64 {
+		result := make([]float64, len(values))
+		for i := range values {
+			result[i] = fun(values[i])
+		}
+		return result
+	}
+}
+
+func transformTimeOffset(values []float64, parameter transformationParameter) []float64 {
```

Comment:
Let's leave out the time shifting for now - I have some other thoughts on handling it.

The biggest issue is that we get bunch of 0s because we don't have enough data. I think the correct way to handle it is by modifying `EvaluationContext` to have a different `TimeRange`. Thus, if you make a (psudocode) call like

```shift(cpu_usage, "2 weeks ago)```

What it would do is - shift doesn't do anything, but changes the evaluation context to fetch 2 weeks earlier data instead. we may need to readjust the returned timerange to ensure that the timeranges are also shifted.